zk.js/src/transaction/solanaTransaction.ts

```diff
@@ -371,22 +408,124 @@ export async function sendAndConfirmShieldedTransaction({
+    systemProof,
     remainingSolanaAccounts,
-    solanaTransactionInputs.transaction,
-    solanaTransactionInputs.pspTransactionInput
-      ? solanaTransactionInputs.pspTransactionInput.verifierIdl
-      : solanaTransactionInputs.transaction.verifierIdl,
-    solanaTransactionInputs.pspTransactionInput,
-    solanaTransactionInputs.pspProof,
-  );
+    accounts,
+    systemPspIdl,
+    pspIdl: pspTransactionInput?.verifierIdl,
+    pspTransactionInput,
+    pspProof,
+    publicTransactionVariables,
+  });
 
   const txHash = await provider.sendAndConfirmShieldedTransaction(instructions);
   const relayerMerkleTreeUpdateResponse = "notPinged";
 
   return { txHash, response: relayerMerkleTreeUpdateResponse };
 }
+
+// TODO: unify event Merkle tree and transaction Merkle tree so that only one is passed
+export function prepareAccounts({
+  transactionAccounts,
+  eventMerkleTreePubkey,
+  signer,
+  relayerRecipientSol,
+  verifierState,
+}: {
+  transactionAccounts: TransactionAccounts;
+  eventMerkleTreePubkey: PublicKey;
+  signer?: PublicKey;
+  relayerRecipientSol?: PublicKey;
+  verifierState?: PublicKey;
+}): lightAccounts {
+  const {
+    senderSol,
+    senderSpl,
+    recipientSol,
+    recipientSpl,
+    relayerPublicKey,
+    pspId,
+    systemPspId,
+  } = transactionAccounts;
+  const verifierProgramId = pspId ? pspId : systemPspId;
+  if (!signer) {
+    signer = relayerPublicKey;
+  }
+  if (!verifierState) {
+    verifierState = PublicKey.findProgramAddressSync(
+      [signer.toBytes(), utils.bytes.utf8.encode("VERIFIER_STATE")],
+      verifierProgramId,
+    )[0];
+  }
+  const accounts = {
+    systemProgramId: SystemProgram.programId,
+    tokenProgram: TOKEN_PROGRAM_ID,
+    logWrapper: SPL_NOOP_PROGRAM_ID,
+    eventMerkleTree: eventMerkleTreePubkey,
+    transactionMerkleTree: transactionAccounts.transactionMerkleTree,
+    registeredVerifierPda: getRegisteredVerifierPda(
+      merkleTreeProgramId,
+      systemPspId,
+    ),
+    authority: getSignerAuthorityPda(merkleTreeProgramId, systemPspId),
+    senderSpl,
+    recipientSpl,
+    senderSol,
+    recipientSol,
+    programMerkleTree: merkleTreeProgramId,
+    tokenAuthority: getTokenAuthority(),
+    verifierProgram: pspId ? systemPspId : undefined,
+    signingAddress: signer,
+    relayerRecipientSol: relayerRecipientSol ? relayerRecipientSol : AUTHORITY,
+    verifierState,
+  };
+  return accounts;
+}
+
+export function getRegisteredVerifierPda(
+  merkleTreeProgramId: PublicKey,
+  verifierProgramId: PublicKey,
+): PublicKey {
+  return PublicKey.findProgramAddressSync(
+    [verifierProgramId.toBytes()],
+    merkleTreeProgramId,
+  )[0];
+}
+
+export function getNullifierPdaPublicKey(
+  nullifier: number[],
+  merkleTreeProgramId: PublicKey,
+) {
+  return PublicKey.findProgramAddressSync(
+    [Uint8Array.from([...nullifier]), utils.bytes.utf8.encode("nf")],
+    merkleTreeProgramId,
+  )[0];
+}
+
+export function getTokenAuthority(): PublicKey {
```

Comment:
naming suggestion for specificity
```suggestion
export function getTokenAuthorityPda(): PublicKey {
```